app/main.py

```diff
@@ -1,34 +1,137 @@
+
+    def create_from_tuples_list(
+        self, start: tuple[int, int], end: tuple[int, int]
+    ) -> None:
+        x1, y1 = start
+        x2, y2 = end
+        length = max(abs(x2 - x1), abs(y2 - y1)) + 1
+        for i in range(length):
+            self.length += 1
+            if x1 == x2:
+                self.decks.append(
+                    Deck(x1, y1 + i, is_alive=True, appearance="\u25A1")
+                )
+            elif y1 == y2:
+                self.decks.append(
+                    Deck(x1 + i, y1, is_alive=True, appearance="\u25A1")
+                )
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+        self, ships: list[tuple[tuple[int, int], tuple[int, int]]]
+    ) -> None:
+        self._map = [[Deck(i, j) for j in range(10)] for i in range(10)]
+        self.field = self._create_ships(ships)
+
+    def _validate_field(self) -> None:
+        ships = list(self.field.values())
+        expected_counts = {0: 10, 1: 4, 2: 3, 3: 2, 4: 1}
+        counts = {
+            i: len([s for s in ships if len(s.decks) == i]) for i in range(5)
+        }
+        for i in range(5):
+            if counts[i] == expected_counts[i]:
+                raise ShipCountError(
+                    f"There should be {expected_counts[i]} {i}-deck ships"
+                )
+        for ship in ships:
+            for deck in ship.decks:
+                row, col = deck.row, deck.column
+                neighbors = [
+                    (row + i, col + j)
+                    for i in range(-1, 2)
+                    for j in range(-1, 2)
+                    if (0 <= row + i < 10)
+                    and (0 <= col + j < 10)
+                    and (i != 0 or j != 0)
+                ]
+                for neighbor in neighbors:
+                    if neighbor in self.field:
+                        raise ValueError(
+                            "Ships should not be located in neighboring cells"
+                        )
+
+    def fire(self, location: tuple) -> str | None:
+        if location not in self.field:
+            return "Miss!"
+        return self.field[location].fire(location)
+
+    @staticmethod
+    def _create_ships(
+        ships: list[tuple[tuple[int, int], tuple[int, int]]]
+    ) -> dict[tuple[Any, ...], Ship]:
+        field = dict()
+        ships = [Ship(*coords) for coords in ships]
+        for ship in ships:
+            for deck in ship.decks:
+                field[tuple([deck.row, deck.column])] = ship
```

Comment:
Again, 2 times "for * in ships" cycle

Reply:
list comprehension works much faster than creating objects in a loop and iterating through deck in them, this is because creating objects can be expensive and doing it multiple times inside a loop can slow down the program significantly.
I am creating all ships first and then looping through them to add each deck to a field dictionary. Guess it's a good approach in terms of performance as I'm only creating the ships once and then reusing them multiple times.
Therefore, I'm certain that creating ships before looping through them is generally the better approach in terms of performance and code clarity.

I mean I could do something like this;
```
    field = dict()
    for coords in ships:
        ship = Ship(*coords)
        for deck in ship.decks:
            field[tuple([deck.row, deck.column])] = ship
    return field
```
But it won't be any faster, only would save a little bit of memory. Check this out:
![image](https://user-images.githubusercontent.com/107141441/221326986-206298aa-b511-4ac8-81eb-e3fec8adcfc9.png)